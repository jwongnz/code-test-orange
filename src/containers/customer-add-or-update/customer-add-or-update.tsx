import React from 'react';
import { Button, TextField } from '@material-ui/core';
import { withRouter, Link, RouteComponentProps } from 'react-router-dom';
import { connect } from 'react-redux';
import { addCustomer, updateCustomer } from '../../store/customers/actions';
import { Customer } from '../../store/customers/types';
import uuid from 'uuid';
import { AppState } from '../../store';
import styled from 'styled-components'

interface MatchParams {
  id: string;
}

interface Props extends RouteComponentProps<MatchParams> {
  customers: Customer[],
  addCustomer: typeof addCustomer,
  updateCustomer: typeof updateCustomer
}

interface State {
  firstName: {
    value: string
    touched: boolean
    valid: boolean
  },
  lastName: {
    value: string
    touched: boolean
    valid: boolean
  },
  dateOfBirth: {
    value: string
    touched: boolean
    valid: boolean
  }
}

const Form = styled.form`
  display: flex;
  flex-direction: column;
  max-width: 400px;
  margin: auto;
`

const ButtonWrapper = styled.div`
  display: flex;
  margin-top: 48px;
`

// TODO: look into loading material UI before styled components
const SaveButton = styled(Button)`
  && {
    margin-left: 20px;
  }
`

export class CustomerAddOrUpdate extends React.Component<Props, State> {
  existingCustomer?: Customer;

  constructor(props: Props) {
    super(props);

    this.existingCustomer = this.props.customers.find((customer) => {
      return customer.id === this.props.match.params.id;
    });
  
    this.state = {
      firstName: {
        value: this.existingCustomer ? this.existingCustomer.firstName : '',
        touched: false,
        valid: this.existingCustomer ? true : false
      },
      lastName: {
        value: this.existingCustomer ? this.existingCustomer.lastName : '',
        touched: false,
        valid: this.existingCustomer ? true : false
      },
      dateOfBirth: {
        value: this.existingCustomer ? this.existingCustomer.dateOfBirth.toISOString().substr(0, 10) : '',
        touched: false,
        valid: this.existingCustomer ? true : false
      }
    };
  }

  handleInputChange = (name: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
    const target = event.target;
    this.setState((current) => ({
      ...current,
      [name]: {
        value: target.value,
        touched: true,
        valid: !!target.value
      }
    }));
  };

  handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.saveCustomer();
  }

  saveCustomer = () => {
    const saveCustomer = this.existingCustomer ? this.props.updateCustomer : this.props.addCustomer;

    saveCustomer({
      id: this.existingCustomer ? this.existingCustomer.id : uuid.v4(),
      firstName: this.state.firstName.value,
      lastName: this.state.lastName.value,
      dateOfBirth: new Date(this.state.dateOfBirth.value)
    });

    this.props.history.push('/');
  }

  formIsValid = () => {
    let valid = true;
    for (let [key, value] of Object.entries(this.state)) {
      if (!value.valid) {
        valid = false;
      }
    }

    return valid;
  }

  render() {
    return (
      <Form noValidate onSubmit={this.handleSubmit}>
        <h1>{this.existingCustomer ? 'Edit' : 'Add'} Customer</h1>
        <TextField
          required
          id="firstName"
          label="First Name"
          value={this.state.firstName.value}
          onChange={this.handleInputChange('firstName')}
          margin="normal"
          error={this.state.firstName.touched && !this.state.firstName.valid}
        />
        <TextField
          required
          id="lastName"
          label="Last Name"
          value={this.state.lastName.value}
          onChange={this.handleInputChange('lastName')}
          margin="normal"
          error={this.state.lastName.touched && !this.state.lastName.valid}
        />
        <TextField
          required
          id="dateOfBirth"
          label="Date of Birth"
          type="date"
          value={this.state.dateOfBirth.value}
          onChange={this.handleInputChange('dateOfBirth')}
          margin="normal"
          InputLabelProps={{
            shrink: true,
          }}
          error={this.state.dateOfBirth.touched && !this.state.dateOfBirth.valid}
        />
        <ButtonWrapper>
          <Button component={Link} to={'/'} variant="contained" color="primary">Cancel</Button>
          <SaveButton type="submit" variant="contained" color="primary" disabled={!this.formIsValid()}>Save</SaveButton>
        </ButtonWrapper>
      </Form>
    )
  };
}

const mapStateToProps = (state: AppState) => ({
  customers: state.customers
});

const mapDispatchToProps = { addCustomer, updateCustomer };

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(CustomerAddOrUpdate));
