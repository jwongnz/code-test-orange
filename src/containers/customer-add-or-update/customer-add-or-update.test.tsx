import React from 'react';
import { CustomerAddOrUpdate } from './customer-add-or-update';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('CustomerAddOrUpdate', () => {
  let props: any;

  beforeEach(() => {
    let mock: any = jest.fn();

    props = {
      customers: [],
      addCustomer: jest.fn(),
      updateCustomer: jest.fn(),
      match: mock,
      location: mock,
      history: {
        push: () => {}
      }
    }
  })

  it('renders correctly', () => {
    const wrapper = shallow(
      <CustomerAddOrUpdate {...props}/>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('enables save button when all inputs are valid', () => {
    // arrange
    const wrapper = shallow<CustomerAddOrUpdate>(
      <CustomerAddOrUpdate {...props}/>
    );
    const instance = wrapper.instance();
    const mockState: any = {
      firstName: {
        valid: true
      },
      lastName: {
        valid: true
      },
      dateOfBirth: {
        valid: true
      }
    };
    instance.setState(mockState);

    // assert
    const saveButton = wrapper.find('[data-test-id="saveButton"]');
    expect(saveButton.props().disabled).toBe(false);
  });

  it('disables save button when an input is invalid', () => {
    // arrange
    const wrapper = shallow<CustomerAddOrUpdate>(
      <CustomerAddOrUpdate {...props}/>
    );
    const instance = wrapper.instance();
    const mockState: any = {
      firstName: {
        valid: true
      },
      lastName: {
        valid: false
      },
      dateOfBirth: {
        valid: true
      }
    };
    instance.setState(mockState);

    // assert
    const saveButton = wrapper.find('[data-test-id="saveButton"]');
    expect(saveButton.props().disabled).toBe(true);
  });

  it('saves customer to store when clicking save', () => {
    // arrange
    const wrapper = shallow<CustomerAddOrUpdate>(
      <CustomerAddOrUpdate {...props}/>
    );
    const instance = wrapper.instance();
    const mockState: any = {
      firstName: {
        valid: true
      },
      lastName: {
        valid: true
      },
      dateOfBirth: {
        valid: true
      }
    };
    instance.setState(mockState);

    // act
    wrapper.find('[data-test-id="form"]').simulate('submit', { preventDefault: () => {} });

    // assert
    expect(props.addCustomer).toHaveBeenCalled();
  });
});