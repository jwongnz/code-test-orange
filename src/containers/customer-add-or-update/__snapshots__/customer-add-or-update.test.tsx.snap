// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerAddOrUpdate renders correctly 1`] = `
<styled.form
  data-test-id="form"
  noValidate={true}
  onSubmit={[Function]}
>
  <h1>
    Add
     Customer
  </h1>
  <WithStyles(ForwardRef(TextField))
    error={false}
    id="firstName"
    label="First Name"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <WithStyles(ForwardRef(TextField))
    error={false}
    id="lastName"
    label="Last Name"
    margin="normal"
    onChange={[Function]}
    required={true}
    value=""
  />
  <WithStyles(ForwardRef(TextField))
    InputLabelProps={
      Object {
        "shrink": true,
      }
    }
    error={false}
    id="dateOfBirth"
    label="Date of Birth"
    margin="normal"
    onChange={[Function]}
    required={true}
    type="date"
    value=""
  />
  <styled.div>
    <WithStyles(ForwardRef(Button))
      color="primary"
      component={[Function]}
      to="/"
      variant="contained"
    >
      Cancel
    </WithStyles(ForwardRef(Button))>
    <Styled(WithStyles(ForwardRef(Button)))
      color="primary"
      data-test-id="saveButton"
      disabled={true}
      type="submit"
      variant="contained"
    >
      Save
    </Styled(WithStyles(ForwardRef(Button)))>
  </styled.div>
</styled.form>
`;
